packages/ui/react-ui-editor/src/extensions/markdown/toolbar.tsx

```diff
@@ -0,0 +1,51 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { type Extension } from '@codemirror/state';
+import { showTooltip, type TooltipView, type EditorView } from '@codemirror/view';
+import React from 'react';
+import { createRoot } from 'react-dom/client';
+
+import { getFormatting } from './formatting';
+import { Toolbar } from '../../components/Toolbar/Toolbar';
+import { useActionHandler } from '../../hooks';
+
+export const toolbar = (): Extension =>
+  showTooltip.compute(['selection'], (state) => {
+    const { main } = state.selection;
+    if (main.empty) {
+      return null;
+    }
+    return {
+      pos: main.from,
+      end: main.to,
+      create: createToolbarTooltip,
+      above: true,
+    };
+  });
+
+const createToolbarTooltip = (view: EditorView): TooltipView => {
+  const wrapper = document.createElement('div');
+  // TODO: Use a portal
```

Comment:
I've been told, in other contexts, that one should use [`createPortal`](https://react.dev/reference/react-dom/createPortal) for this, rather than `createRoot`, since multiple React roots are expensive or otherwise problematic. But that requires the portal to somehow connect to some higher-up React tree, and I'm not entirely sure how to set that up.

Reply:
I'll confer with the the team's React expert, but note that we do this elsewhere in the codebase by just supplying `document.body` as the parent div.

Reply:
Moot here anyway if we're not doing this as a tooltip over the editor.